### ISSUE TITLE
有关detection subnet training 生成anchor速度的问题

### ISSUE THREAD

**Issue by accountcwd** [LiMeng95/MultiPoseNet.pytorch#6]
你好，感谢你们提供的工程，我在训练检测子网络的时候发现训练速度是keypoints的1/6（keypoints 30fps,detection 5fps），打点发现生成 anchors非常费时间，分类和回归只需要2ms，而anchors却要200ms。
定位到最后发现，生成anchor是在cpu上进行，
最后执行
`torch.from_numpy(all_anchors.astype(np.float32)).cuda()`
转到GPU内存时，花了200ms。我对pytorch不是很熟，请问anchor这个操作能直接在gpu上创建并计算实现吗？

**Comment by LiMeng95** [comment]
@accountcwd , 你好，
detection subnet这部分我目前是整合了[这个项目](https://github.com/yhenon/pytorch-retinanet)的训练和测试过程。
关于生成anchors的问题，这部分理论上来说是可以直接通过tensor在GPU上生成和操作的，不过这里涉及了很多numpy的计算过程，可能难以完全在GPU上计算。另外还需要考虑的问题是，这部分有很多中间变量，如果直接在GPU上计算的话，会占用很多GPU资源，可能会产生显存不足的问题。
关于detection subnet这部分的训练和测试过程，目前确实存在一些问题。之后计划进行改进。

感谢对这个项目的关注，欢迎提出更多意见和建议~

**Comment by accountcwd** [comment]
@LiMeng95 感谢你的回答，这两天我尝试过将Numpy的操作转移到GPU上，发现torch.cuda.FloatTensor有好多操作实现不了(也有可能是我水平不够)，的确还有你所说的内存问题。
所以我在想先把这部分操作预处理了，跟图片一起直接load进gpu，不知你觉得这个方法怎样呢？

**Comment by LiMeng95** [comment]
@accountcwd ，你好
你说的一起load的思路是可行的，但是一定需要很多工作量，因为不仅是在加载数据这部分需要修改，另外还要修改data augmentation部分的代码。并且会导致数据加载过程变慢。这样考虑的话，将anchor直接load进GPU的思路并不一定能显著减小训练时间。
我认为现在detection subnet部分训练过程中最主要的问题并不是在于生成anchors这部分的时间消耗，而是在于GPU计算分布不均。如果采用多GPU的话，会发现第一个GPU占用显存会远高于其他GPU，目前我认为是计算focal loss这部分存在一些问题，正在想办法改进。

**Comment by accountcwd** [comment]
@LiMeng95 谢谢你的详细回答，我当时没考虑到还有数据增强要处理，这样的话估计会得不偿失。
然后我详细分析了训练的细节：
我这边环境是单显卡GTX1080，8G显存，以下是逐条语句测试运行时间的结果：
batch size 为4，一个batch运行时间是 #**750ms**
其中：
self.model(*inputs) #**220ms**
self.model.module.build_loss(saved_for_loss, *gts) #**240ms**
loss.backward()  #**220ms**

在 forward中主要运算时间集中在anchor中的语句：
`torch.from_numpy(all_anchors.astype(np.float32)).cuda() #200ms`

而第二段，时间集中在函数FocalLoss.forward的**各种.cuda()的操作**中，加起来也差不多200多ms了。
第三段backward中，不知道具体反传的机制是不是有CPU与GPU的交互。

我的观察是训练大部分的时间花在了内存之间的传输中，如果能减少这些传输操作应该就能提高速度了。希望能给你提供解决这个问题的灵感，我也尝试尽我能力修改一些语句试试试。

**Comment by LiMeng95** [comment]
@accountcwd ，感谢提供detection subnet训练过程中的详细时间信息

**Comment by accountcwd** [comment]
你好，对于这个问题我有新的想法，这个Anchors每一次[forward](https://github.com/LiMeng95/MultiPoseNet.pytorch/blob/7b548fc20ae5b0141a2113669e8606103fb7e6d4/network/anchors.py#L21)都是生成同样的东西(只根据输入图片的尺寸生成)，那为什么不可以改在初始化的时候预处理好塞进显存，然后forward的时候直接读取呢？按照我上面测的结果，这应该能减少约1/3的时间。

**Comment by LiMeng95** [comment]
是的，生成Anchor这部分仅用了图像尺寸。当时这样写是为了处理图像大小不同的情况。

目前来说，在固定输入大小的情况下，是可以直接读取anchor数据的，不需要每次都计算anchor尺寸。
